File: kiko-api/src/types/judgeTypes.ts
/**
 * Judge System Types
 * Launchpad Decision Engine v3.5
 * 
 * Based on Juage.md specification
 */

// ========================
// INPUT TYPES
// ========================

export interface DecisionEngineInput {
    user_amount: number;        // USD
    token_address: string;
    launchpad_type: string;     // e.g., 'pump.fun', 'zora', 'four.meme'
    chain: string;
    chain_id: number;
    timestamp: string;
    target_wallet?: string;     // For copy trade context
}

// ========================
// USER SIZE LAYER
// ========================

export type UserSizeLevel = 'L1' | 'L2' | 'L3' | 'L4';

export interface UserSizeLayer {
    user_size_level: UserSizeLevel;
    max_slippage_allowed: number;   // e.g., 0.05 = 5%
    max_risk_allowed: number;       // 0-1
    score: number;                  // 0-1 (higher = safer for user)
    reasons: string[];
}

// User Size Thresholds (USD)
export const USER_SIZE_THRESHOLDS = {
    L1: { max: 100, label: 'Small Probe' },
    L2: { min: 100, max: 400, label: 'Medium' },
    L3: { min: 400, max: 1500, label: 'Large' },
    L4: { min: 1500, label: 'Heavy Position' },
} as const;

// ========================
// LIQUIDITY LAYER
// ========================

export type LayerDecision = 'ALLOW' | 'ALLOW_WITH_RISK' | 'BLOCK';

export interface LiquidityLayer {
    lp_depth_usd: number;
    user_amount_usd: number;
    slippage_estimate: number;       // 0-1 (0.05 = 5%)
    liquidity_risk_score: number;    // 0-1 (higher = safer)
    liquidity_decision: LayerDecision;
    reasons: string[];
}

// ========================
// STRUCTURE LAYER
// ========================

export interface StructureFeatures {
    has_bonding_curve: boolean;
    has_fixed_pool: boolean;
    has_migration: boolean;
    creator_fee: number;            // 0-1 (0.05 = 5%)
    curve_type: string;             // 'linear', 'exponential', 'fixed', 'unknown'
    lp_lock_info: string;           // 'locked', 'unlocked', 'unknown'
    metadata_quality: number;       // 0-1
}

export interface StructureLayer {
    launchpad_type: string;
    structure_features: StructureFeatures;
    structure_risk_score: number;   // 0-1 (higher = safer)
    structure_decision: LayerDecision;
    reasons: string[];
}

// ========================
// STAGE LAYER
// ========================

export type TokenStage = 'S0' | 'S1' | 'S2' | 'S3' | 'S4';

export interface StageLayer {
    stage: TokenStage;
    contract_age_minutes: number;
    stage_risk_score: number;       // 0-1 (higher = safer)
    stage_decision: LayerDecision;
    reasons: string[];
}

// Stage Thresholds (minutes)
export const STAGE_THRESHOLDS = {
    S0: { max: 10, label: 'Just Created', risk: 0.2 },
    S1: { min: 10, max: 120, label: 'Early', risk: 0.4 },
    S2: { min: 120, max: 1440, label: 'Mid-term', risk: 0.6 },
    S3: { min: 1440, label: 'Stable', risk: 0.8 },
    S4: { label: 'Post-Migration / Special', risk: 0.5 },
} as const;

// ========================
// TOKEN INTELLIGENCE LAYER
// ========================

export interface ProjectIdentity {
    has_official_twitter: boolean;
    has_team_identity: boolean;
    has_logo: boolean;
    has_description: boolean;
    contract_verified: boolean;
    contract_age_hours: number;
    launchpad_metadata_quality: number;  // 0-1
    score: number;                       // 0-1
}

export interface SocialPresence {
    twitter_active: boolean;
    twitter_followers: number;
    has_kol_mentions: boolean;
    community_discussion_level: number;  // 0-1
    telegram_active: boolean;
    discord_active: boolean;
    score: number;                       // 0-1
}

export interface NarrativeStrength {
    narrative_type: string;              // 'meme', 'AI', 'infra', 'culture', 'celebrity', 'unknown'
    narrative_strength: number;          // 0-1
    narrative_alignment: number;         // 0-1 (market trend alignment)
    narrative_consistency: number;       // 0-1 (cross-channel consistency)
    score: number;                       // 0-1
}

export interface WebsiteQuality {
    has_website: boolean;
    website_reachable: boolean;
    website_design_quality: number;      // 0-1
    website_content_depth: number;       // 0-1
    has_docs: boolean;
    has_product: boolean;
    score: number;                       // 0-1
}

export interface RiskSignals {
    scam_reports_found: boolean;
    rug_reports_found: boolean;
    honeypot_reports_found: boolean;
    negative_sentiment_score: number;    // 0-1 (higher = more negative)
    deployer_reputation: number;         // 0-1 (higher = better)
    score: number;                       // 0-1 (safety score, higher = safer)
}

export type TokenIntelligenceLabel = 'strong' | 'medium' | 'weak' | 'danger';

export interface TokenIntelligenceLayer {
    project_identity: ProjectIdentity;
    social_presence: SocialPresence;
    narrative_strength: NarrativeStrength;
    website_quality: WebsiteQuality;
    risk_signals: RiskSignals;

    token_intelligence_score: number;    // 0-1 (aggregate)
    token_intelligence_label: TokenIntelligenceLabel;
    risk_tags: string[];                 // e.g., ['no-docs', 'strong-social', 'meme']
    reasons: string[];
}

// ========================
// FINAL DECISION
// ========================

export type FinalDecision = 'ALLOW' | 'ALLOW_WITH_RISK' | 'BLOCK';

export interface FinalDecisionOutput {
    decision: FinalDecision;
    overall_risk_score: number;          // 0-1 (higher = safer)
    slippage_estimate: number;           // 0-1
    reasons: string[];
    ai_rationale?: string;
}

// ========================
// COMPLETE OUTPUT
// ========================

export interface DecisionEngineLayers {
    user_size_layer: UserSizeLayer;
    liquidity_layer: LiquidityLayer;
    structure_layer: StructureLayer;
    stage_layer: StageLayer;
    token_intelligence_layer: TokenIntelligenceLayer;
}

export interface DecisionEngineOutput {
    decision_engine: {
        input: DecisionEngineInput;
        layers: DecisionEngineLayers;
        final_decision: FinalDecisionOutput;
        decision_id?: string;
    };
}

// ========================
// HELPER TYPES
// ========================

export interface TokenData {
    address: string;
    symbol: string;
    name: string;
    price: number;
    liquidity: number;
    fdv: number;
    marketCap: number;
    priceChange5m: number;
    priceChange1h: number;
    priceChange24h: number;
    pairCreatedAt?: number;
    poolAddress?: string;
    socials?: Array<{ type: string; url: string }>;
    websites?: Array<{ url: string }>;
}

export interface SecurityData {
    status: 'Unknown' | 'Safe' | 'Warning' | 'High Risk' | 'Critical' | 'Medium';
    riskScore: number;
    isHoneypot: boolean;
    isMintable: boolean;
    isProxy: boolean;
    buyTax: number;
    sellTax: number;
    warnings: string[];
    positives: string[];
    recommendation: string;
    details: {
        isOpenSource: boolean;
        hasRenouncedOwner: boolean;
        isMintable: boolean;
        canDisableTrade: boolean;
        isBlacklisted: boolean;
    };
    source: string;
    lpLocked: boolean;
}

File: kikoapi/internal/types/judge_types.go
// Package types: Judge System Types (Launchpad Decision Engine).
// Migrated from kiko-api/src/types/judgeTypes.ts

package types

// DecisionEngineInput is the input to the decision engine.
type DecisionEngineInput struct {
	UserAmount    float64 `json:"user_amount"`
	TokenAddress  string  `json:"token_address"`
	LaunchpadType string  `json:"launchpad_type"`
	Chain         string  `json:"chain"`
	ChainID       int     `json:"chain_id"`
	Timestamp     string  `json:"timestamp"`
	TargetWallet  string  `json:"target_wallet,omitempty"`
}

// UserSizeLevel represents user size tier.
type UserSizeLevel string

const (
	UserSizeL1 UserSizeLevel = "L1"
	UserSizeL2 UserSizeLevel = "L2"
	UserSizeL3 UserSizeLevel = "L3"
	UserSizeL4 UserSizeLevel = "L4"
)

// UserSizeLayer holds user size layer output.
type UserSizeLayer struct {
	UserSizeLevel     UserSizeLevel `json:"user_size_level"`
	MaxSlippageAllowed float64      `json:"max_slippage_allowed"`
	MaxRiskAllowed    float64      `json:"max_risk_allowed"`
	Score             float64      `json:"score"`
	Reasons           []string    `json:"reasons"`
}

// UserSizeThreshold defines min/max and label for a tier.
type UserSizeThreshold struct {
	Min   float64 `json:"min,omitempty"`
	Max   float64 `json:"max,omitempty"`
	Label string  `json:"label"`
}

// USER_SIZE_THRESHOLDS holds tier thresholds (USD).
var USER_SIZE_THRESHOLDS = map[UserSizeLevel]UserSizeThreshold{
	UserSizeL1: {Max: 100, Label: "Small Probe"},
	UserSizeL2: {Min: 100, Max: 400, Label: "Medium"},
	UserSizeL3: {Min: 400, Max: 1500, Label: "Large"},
	UserSizeL4: {Min: 1500, Label: "Heavy Position"},
}

// LayerDecision is the decision from a layer.
type LayerDecision string

const (
	LayerAllow        LayerDecision = "ALLOW"
	LayerAllowWithRisk LayerDecision = "ALLOW_WITH_RISK"
	LayerBlock        LayerDecision = "BLOCK"
)

// LiquidityLayer holds liquidity layer output.
type LiquidityLayer struct {
	LpDepthUSD         float64        `json:"lp_depth_usd"`
	UserAmountUSD      float64        `json:"user_amount_usd"`
	SlippageEstimate   float64        `json:"slippage_estimate"`
	LiquidityRiskScore float64        `json:"liquidity_risk_score"`
	LiquidityDecision LayerDecision  `json:"liquidity_decision"`
	Reasons            []string       `json:"reasons"`
}

// StructureFeatures describes launchpad structure.
type StructureFeatures struct {
	HasBondingCurve   bool    `json:"has_bonding_curve"`
	HasFixedPool      bool    `json:"has_fixed_pool"`
	HasMigration      bool    `json:"has_migration"`
	CreatorFee        float64 `json:"creator_fee"`
	CurveType         string  `json:"curve_type"`
	LpLockInfo        string  `json:"lp_lock_info"`
	MetadataQuality   float64 `json:"metadata_quality"`
}

// StructureLayer holds structure layer output.
type StructureLayer struct {
	LaunchpadType     string           `json:"launchpad_type"`
	StructureFeatures StructureFeatures `json:"structure_features"`
	StructureRiskScore float64         `json:"structure_risk_score"`
	StructureDecision LayerDecision   `json:"structure_decision"`
	Reasons           []string        `json:"reasons"`
}

// TokenStage represents token lifecycle stage.
type TokenStage string

const (
	StageS0 TokenStage = "S0"
	StageS1 TokenStage = "S1"
	StageS2 TokenStage = "S2"
	StageS3 TokenStage = "S3"
	StageS4 TokenStage = "S4"
)

// StageLayer holds stage layer output.
type StageLayer struct {
	Stage           TokenStage   `json:"stage"`
	ContractAgeMins int          `json:"contract_age_minutes"`
	StageRiskScore  float64     `json:"stage_risk_score"`
	StageDecision   LayerDecision `json:"stage_decision"`
	Reasons         []string    `json:"reasons"`
}

// StageThreshold defines stage thresholds.
type StageThreshold struct {
	Min   int     `json:"min,omitempty"`
	Max   int     `json:"max,omitempty"`
	Label string  `json:"label"`
	Risk  float64 `json:"risk,omitempty"`
}

// STAGE_THRESHOLDS holds stage thresholds (minutes).
var STAGE_THRESHOLDS = map[TokenStage]StageThreshold{
	StageS0: {Max: 10, Label: "Just Created", Risk: 0.2},
	StageS1: {Min: 10, Max: 120, Label: "Early", Risk: 0.4},
	StageS2: {Min: 120, Max: 1440, Label: "Mid-term", Risk: 0.6},
	StageS3: {Min: 1440, Label: "Stable", Risk: 0.8},
	StageS4: {Label: "Post-Migration / Special", Risk: 0.5},
}

// ProjectIdentity holds project identity metrics.
type ProjectIdentity struct {
	HasOfficialTwitter       bool    `json:"has_official_twitter"`
	HasTeamIdentity          bool    `json:"has_team_identity"`
	HasLogo                  bool    `json:"has_logo"`
	HasDescription           bool    `json:"has_description"`
	ContractVerified         bool    `json:"contract_verified"`
	ContractAgeHours         float64 `json:"contract_age_hours"`
	LaunchpadMetadataQuality float64 `json:"launchpad_metadata_quality"`
	Score                    float64 `json:"score"`
}

// SocialPresence holds social presence metrics.
type SocialPresence struct {
	TwitterActive             bool    `json:"twitter_active"`
	TwitterFollowers          int     `json:"twitter_followers"`
	HasKolMentions            bool    `json:"has_kol_mentions"`
	CommunityDiscussionLevel  float64 `json:"community_discussion_level"`
	TelegramActive            bool    `json:"telegram_active"`
	DiscordActive             bool    `json:"discord_active"`
	Score                     float64 `json:"score"`
}

// NarrativeStrength holds narrative metrics.
type NarrativeStrength struct {
	NarrativeType        string  `json:"narrative_type"`
	NarrativeStrength    float64 `json:"narrative_strength"`
	NarrativeAlignment   float64 `json:"narrative_alignment"`
	NarrativeConsistency float64 `json:"narrative_consistency"`
	Score                float64 `json:"score"`
}

// WebsiteQuality holds website quality metrics.
type WebsiteQuality struct {
	HasWebsite          bool    `json:"has_website"`
	WebsiteReachable    bool    `json:"website_reachable"`
	WebsiteDesignQuality float64 `json:"website_design_quality"`
	WebsiteContentDepth float64 `json:"website_content_depth"`
	HasDocs             bool    `json:"has_docs"`
	HasProduct          bool    `json:"has_product"`
	Score               float64 `json:"score"`
}

// RiskSignals holds risk signal metrics.
type RiskSignals struct {
	ScamReportsFound     bool    `json:"scam_reports_found"`
	RugReportsFound      bool    `json:"rug_reports_found"`
	HoneypotReportsFound bool    `json:"honeypot_reports_found"`
	NegativeSentimentScore float64 `json:"negative_sentiment_score"`
	DeployerReputation   float64 `json:"deployer_reputation"`
	Score                float64 `json:"score"`
}

// TokenIntelligenceLabel is the aggregate label.
type TokenIntelligenceLabel string

const (
	TokenIntelligenceStrong TokenIntelligenceLabel = "strong"
	TokenIntelligenceMedium TokenIntelligenceLabel = "medium"
	TokenIntelligenceWeak   TokenIntelligenceLabel = "weak"
	TokenIntelligenceDanger TokenIntelligenceLabel = "danger"
)

// TokenIntelligenceLayer holds token intelligence layer output.
type TokenIntelligenceLayer struct {
	ProjectIdentity       ProjectIdentity   `json:"project_identity"`
	SocialPresence        SocialPresence   `json:"social_presence"`
	NarrativeStrength     NarrativeStrength `json:"narrative_strength"`
	WebsiteQuality        WebsiteQuality   `json:"website_quality"`
	RiskSignals           RiskSignals      `json:"risk_signals"`
	TokenIntelligenceScore  float64         `json:"token_intelligence_score"`
	TokenIntelligenceLabel TokenIntelligenceLabel `json:"token_intelligence_label"`
	RiskTags              []string         `json:"risk_tags"`
	Reasons               []string         `json:"reasons"`
}

// FinalDecision is the final decision type.
type FinalDecision string

const (
	FinalAllow         FinalDecision = "ALLOW"
	FinalAllowWithRisk FinalDecision = "ALLOW_WITH_RISK"
	FinalBlock         FinalDecision = "BLOCK"
)

// FinalDecisionOutput holds the final decision output.
type FinalDecisionOutput struct {
	Decision           FinalDecision `json:"decision"`
	OverallRiskScore   float64       `json:"overall_risk_score"`
	OverallRiskLevel   string        `json:"overall_risk_level,omitempty"`
	SlippageEstimate   float64       `json:"slippage_estimate"`
	Reasons            []string      `json:"reasons"`
	AiRationale        string        `json:"ai_rationale,omitempty"`
}

// DecisionEngineLayers aggregates all layer outputs.
type DecisionEngineLayers struct {
	UserSizeLayer         UserSizeLayer         `json:"user_size_layer"`
	LiquidityLayer        LiquidityLayer        `json:"liquidity_layer"`
	StructureLayer        StructureLayer        `json:"structure_layer"`
	StageLayer            StageLayer            `json:"stage_layer"`
	TokenIntelligenceLayer TokenIntelligenceLayer `json:"token_intelligence_layer"`
}

// DecisionEngineOutput is the full decision engine response.
type DecisionEngineOutput struct {
	DecisionEngine struct {
		Input          DecisionEngineInput  `json:"input"`
		Layers         DecisionEngineLayers `json:"layers"`
		FinalDecision FinalDecisionOutput  `json:"final_decision"`
		DecisionID    string               `json:"decision_id,omitempty"`
	} `json:"decision_engine"`
}

// TokenData is helper type for token data.
type TokenData struct {
	Address        string  `json:"address"`
	Symbol         string  `json:"symbol"`
	Name           string  `json:"name"`
	Price          float64 `json:"price"`
	Liquidity      float64 `json:"liquidity"`
	Fdv            float64 `json:"fdv"`
	MarketCap      float64 `json:"marketCap"`
	PriceChange5m  float64 `json:"priceChange5m"`
	PriceChange1h  float64 `json:"priceChange1h"`
	PriceChange24h float64 `json:"priceChange24h"`
	PairCreatedAt  *int64  `json:"pairCreatedAt,omitempty"`
	PoolAddress    string  `json:"poolAddress,omitempty"`
	Socials        []struct {
		Type string `json:"type"`
		URL  string `json:"url"`
	} `json:"socials,omitempty"`
	Websites []struct {
		URL string `json:"url"`
	} `json:"websites,omitempty"`
}

// SecurityDataStatus is the status string for SecurityData.
type SecurityDataStatus string

const (
	SecurityStatusUnknown  SecurityDataStatus = "Unknown"
	SecurityStatusSafe     SecurityDataStatus = "Safe"
	SecurityStatusWarning  SecurityDataStatus = "Warning"
	SecurityStatusHighRisk SecurityDataStatus = "High Risk"
	SecurityStatusCritical SecurityDataStatus = "Critical"
	SecurityStatusMedium   SecurityDataStatus = "Medium"
)

// SecurityDataDetails holds security detail flags.
type SecurityDataDetails struct {
	IsOpenSource      bool `json:"isOpenSource"`
	HasRenouncedOwner bool `json:"hasRenouncedOwner"`
	IsMintable        bool `json:"isMintable"`
	CanDisableTrade   bool `json:"canDisableTrade"`
	IsBlacklisted     bool `json:"isBlacklisted"`
}

// SecurityData is helper type for token security.
type SecurityData struct {
	Status         SecurityDataStatus  `json:"status"`
	RiskScore      float64             `json:"riskScore"`
	IsHoneypot     bool                `json:"isHoneypot"`
	IsMintable     bool                `json:"isMintable"`
	IsProxy        bool                `json:"isProxy"`
	BuyTax         float64             `json:"buyTax"`
	SellTax        float64             `json:"sellTax"`
	Warnings       []string            `json:"warnings"`
	Positives      []string            `json:"positives"`
	Recommendation string              `json:"recommendation"`
	Details        SecurityDataDetails `json:"details"`
	Source         string              `json:"source"`
	LpLocked       bool                `json:"lpLocked"`
}
